src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs

```diff
@@ -0,0 +1,200 @@
+        /// </remarks>
+        public int Order => -1000 + 100;
+
+        public void OnProvidersExecuted(ApplicationModelProviderContext context)
+        {
+        }
+
+        public void OnProvidersExecuting(ApplicationModelProviderContext context)
+        {
+            foreach (var controllerModel in context.Result.Controllers)
+            {
+                var isApiController = controllerModel.Attributes.OfType<IApiBehaviorMetadata>().Any();
+                var controllerHasSelectorModel = controllerModel.Selectors.Any(s => s.AttributeRouteModel != null);
+
+                foreach (var actionModel in controllerModel.Actions)
+                {
+                    if (!isApiController && !actionModel.Attributes.OfType<IApiBehaviorMetadata>().Any())
+                    {
+                        continue;
+                    }
+
+                    EnsureActionIsAttributeRouted(controllerHasSelectorModel, actionModel);
+
+                    AddInvalidModelStateFilter(actionModel);
+
+                    InferParameterBindingSources(actionModel);
+                }
+            }
+        }
+
+        private static void EnsureActionIsAttributeRouted(bool controllerHasSelectorModel, ActionModel actionModel)
+        {
+            if (!controllerHasSelectorModel && !actionModel.Selectors.Any(s => s.AttributeRouteModel != null))
+            {
+                // Require attribute routing with controllers annotated with ApiControllerAttribute
+                throw new InvalidOperationException(Resources.FormatApiController_AttributeRouteRequired(nameof(ApiControllerAttribute)));
+            }
+        }
+
+        private void AddInvalidModelStateFilter(ActionModel actionModel)
+        {
+            if (_apiBehaviorOptions.SuppressModelStateInvalidFilter)
+            {
+                return;
+            }
+
+            Debug.Assert(_apiBehaviorOptions.InvalidModelStateResponseFactory != null);
+            actionModel.Filters.Add(_modelStateInvalidFilter);
+        }
+
+        private void InferParameterBindingSources(ActionModel actionModel)
+        {
+            if (_modelMetadataProvider == null || _apiBehaviorOptions.SuppressInferBindingSourcesForParameters)
+            {
+                return;
+            }
+            var inferredBindingSources = new BindingSource[actionModel.Parameters.Count];
+            var foundFromBodyParameter = false;
+
+            for (var i = 0; i < inferredBindingSources.Length; i++)
+            {
+                var parameter = actionModel.Parameters[i];
+                var bindingSource = InferBindingSourceForParameter(parameter);
+                if (bindingSource == BindingSource.Body)
+                {
+                    if (foundFromBodyParameter)
+                    {
+                        // More than one parameter is inferred as FromBody. Log a warning and skip this action.
+                        _logger.UnableToInferBindingSource(actionModel);
```

Comment:
Can you log an issue to decide if this should be an error instead? (throwing)

I don't want to block you on this.